### PR TITLE
refactor: Add type constraints to internal op definitions.

### PR DESCRIPTION
**Pull request by TrevorBergeron**
Thank you for opening a Pull Request! Before submitting your PR, there are a few things you can do to make sure it goes smoothly:
- [ ] Make sure to open an issue as a [bug/issue](https://github.com/googleapis/python-bigquery-dataframes/issues/new/choose) before writing your code!  That way we can discuss the change, evaluate designs, and agree on the general idea
- [ ] Ensure the tests and linter pass
- [ ] Code coverage does not decrease (if any source code was changed)
- [ ] Appropriate docs were updated (if necessary)

Fixes #<issue_number_goes_here> 🦕

### CODE REVIEW

File: bigframes/operations/type.py [hunk 1]
```diff
@@ -22,54 +23,175 @@
 
 
 @dataclasses.dataclass
-class OpTypeRule:
-    def output_type(self, *input_types: ExpressionType) -> ExpressionType:
-        raise NotImplementedError("Abstract typing rule has no output type")
+class TypeSignature(abc.ABC):
```
Comment (chelsea-lin) [inline bigframes/operations/type.py:26]: Please add a docstring and give some context here?

File: bigframes/operations/type.py [hunk 2]
```diff
@@ -22,54 +23,175 @@
...
+        lcd_type = bigframes.dtypes.lcd_etype(left_type, right_type)
+        if lcd_type == bigframes.dtypes.INT_DTYPE:
+            return bigframes.dtypes.FLOAT_DTYPE
+        return lcd_type
 
 
 @dataclasses.dataclass
-class Fixed(OpTypeRule):
-    out_type: ExpressionType
+class Numeric(BinaryTypeSignature):
```
Comment (chelsea-lin) [inline bigframes/operations/type.py:126]: `BinaryNumeric`? Why we have both this and `BinaryRealNumeric`?

File: bigframes/operations/type.py [hunk 3]
```diff
@@ -22,54 +23,175 @@
...
+    def output_type(
+        self, left_type: ExpressionType, right_type: ExpressionType
+    ) -> ExpressionType:
+        if not bigframes.dtypes.is_numeric(left_type):
+            raise TypeError(f"Type {left_type} is not numeric")
+        if not bigframes.dtypes.is_numeric(right_type):
+            raise TypeError(f"Type {right_type} is not numeric")
+        lcd_type = bigframes.dtypes.lcd_etype(left_type, right_type)
+        if lcd_type == bigframes.dtypes.INT_DTYPE:
+            return bigframes.dtypes.FLOAT_DTYPE
```
Comment (chelsea-lin) [inline bigframes/operations/type.py:121]: please add a comment to explain why we need a special case here?

File: bigframes/operations/__init__.py [hunk 4]
```diff
@@ -392,9 +458,15 @@ class StructFieldOp(UnaryOp):
     name_or_index: str | int
 
     def output_type(self, *input_types):
-        pd_type = typing.cast(pd.ArrowDtype, input_types[0])
-        pa_struct_t = typing.cast(pa.StructType, pd_type.pyarrow_dtype)
-        pa_result_type = pa_struct_t[self.name_or_index].type
+        input_type = input_types[0]
+        if not isinstance(input_type, pd.ArrowDtype):
+            raise TypeError("field accessor input must be a struct type")
```
Comment (chelsea-lin) [inline bigframes/operations/__init__.py:463]: typo: array?

**Review by chelsea-lin — CHANGES_REQUESTED** [review]

File: bigframes/operations/type.py (hunk 2 above)
Comment (TrevorBergeron) [reply bigframes/operations/type.py:126]: The difference is whether integer inputs result in integer output for a float output. Some operations, such as logarithm, will produce float result even if input is an integer eg `log(3) ~= 0.47712`. Other operations, such as abs, preserve int type eg `abs(-2) == 2`

File: bigframes/operations/type.py (hunk 3 above)
Comment (TrevorBergeron) [reply bigframes/operations/type.py:121]: Added comment

File: bigframes/operations/__init__.py (hunk 4 above)
Comment (TrevorBergeron) [reply bigframes/operations/__init__.py:463]: this is for structs

File: bigframes/operations/type.py (hunk 3 above)
Comment (chelsea-lin) [reply bigframes/operations/type.py:128]: Thanks! It would be much easier to follow then!

**Review by chelsea-lin — APPROVED** [review]
LGTM overall. If the change grows after fixing the presubmit tests, I can take another look.

**Pull request merged**